### PR TITLE
Add global optimization to limit redundant work

### PR DESCRIPTION
**Pull request by rjzamora**
Rough exploration of a possible solution to #218

This PR is very far from being "ready for review", but I wanted to start experimenting in public. I don't like the current naming, API, or organization yet.

### CODE REVIEW

**Comment by rjzamora** [comment]
@phofl @mrocklin - I'd be interested to get feedback on the general direction of this PR.

As far as I can tell, the logic in this PR should address the problematic examples described in #218. The basic idea is that we add a new optimization pass within `Expr.optimize` - *after* the `simplify` step and *before* the `optimize_blockwise_fusion` step. This new `simplify_global` optimization pass is similar to `Expr.simplify`, except (1) we pass along a reference to the final "root" expression, and (2) we always rewrite/modify children **before** parents. While (1) allows the local `_simplify_global` logic to compare the current `Expr` object with similar nodes in the global graph, (2) makes it easier for us to effectively "push back up" problematic `Projection` operations.

What I'm not sure about:

- Are there other (better) ways to get the same result? If not...
- What should we call this optimization pass? I used `simplify_global` for now to be as general as possible, but I cannot say that I'm a fan (next bullet point is related).
- Should the code for this optimization pass be more specific and isolated? I originally assumed that the `simplify_global` "control flow" could be used for other optimizations that need access to `root`. However, I am now thinking that we should be more explicit that the purpose of this logic is to "pull back" problematic column-projection state.

File: dask_expr/io/parquet.py [hunk 1]
```diff
@@ -427,6 +427,46 @@ def columns(self):
...
+        # access by aggregating multiple operations with different
+        # column projections into the same operation.
+        alike = self._find_similar_operations(root, ignore=["columns", "_series"])
+        if alike:
+            # We have other ReadParquet operations in the expression
+            # graph that can be combined with this one, because
+            # the only difference is the column projection
+
+            # Find the column-projection union needed to combine
+            # the qulified ReadParquet operations
```
Comment (phofl) [inline dask_expr/io/parquet.py:441]: typo

**Review by phofl — COMMENTED** [review]
This generally looks good to me. I think I agree with your assessment that we should limit this particular step to combining similar expressions for now. Maybe something like this as the name?

**Comment by rjzamora** [comment]
>I think I agree with your assessment that we should limit this particular step to combining similar expressions for now. Maybe something like this as the name?

How about `combine_similar`?

**Comment by phofl** [comment]
Sounds good!

**Review by phofl — APPROVED** [review]

**Comment by phofl** [comment]
Ready to merge from your PoV?

**Comment by rjzamora** [comment]
Thanks for the review @phofl !

>Ready to merge from your PoV?

I feel that the actual optimization logic is fine for now. It is completely separate from `simplify`, and can always be improved/removed as needed.

My only concern is that we may want to add some simple tests for the new `_substitute_parameters` and `_find_similar_operations` methods. Also, I wonder if those "methods" should be moved over to `_utils` for now? Wdyt?

**Comment by phofl** [comment]
Yeah I think moving them over makes sense.
Not against tests, but I think this would be fine as is too

**Comment by rjzamora** [comment]
>Yeah I think moving them over makes sense.
Not against tests, but I think this would be fine as is too

Okay - I think I'll merge so that this is out of the way.

**Pull request merged**